dora-parser/src/interner.rs

```diff
@@ -46,20 +46,49 @@ impl Deref for ArcStr {
 
 pub struct Interner {
     data: Mutex<Internal>,
+    pub known_names: KnownNames,
```

Comment:
If I remember correctly, you suggested that we shouldn't have to lookup these rather frequent names over and over ...

Reply:
Yep, but couldn't we store it in e.g. src/vm/vip.rs? It already knows about other "special" classes/functions, might make sense to store it there.

Reply:
Should be possible, yes!

Reply:
With InternalAnnotation we might not need to track names of annotations at all.

Reply:
Why would this be harder? This would be pretty similar to functions where we have FctKind::Builtin?

Reply:
You would need some code to wire up the internal annotations that you wouldn't need otherwise, compared to interning the names which is optional.

Reply:
Plus, how to determine whether and how something needs to be wired up?
And if we hard-wiring things anyway, then the `InternalAnnotation` elements could carry the methods to access individual parameter values (e. g. `message`, `since` for `@deprecated`).

Reply:
It would be done in prelude.rs like all the other intrinsics/functions/classes known to the compiler.

Reply:
And if we register the whole annotation in prelude, then having them as source files becomes rather pointless.

Reply:
Ideally, I wouldn't want to have the `internal` field on Annotation at all, neither as `bool` nor as `InternalAnnotation`.

Reply:
Ok, here is my current understanding:

1. Annotations like `@open/@override/@final` etc. need special logic in the compiler. For example `@override` can only be used on a method, which actually overrides a method of a parent class and that method can't be `@final`. So my assumption is that there needs to be some way to recognize these handles in the runtime.

> Ideally, I wouldn't want to have the internal field on Annotation at all, neither as bool nor as InternalAnnotation.
It just seems that this unavoidable. Even some future version of Dora couldn't get rid of that.

2. If we need a way to recognize at least some internal or special annotations, this InternalAnnotation enum seems more robust than using the Name returned by the interner. First, InternalAnnotation will definitely not break when we introduce namespace/package/module (or whatever it is called then) and the whole standard library is moved into the `stdlib` package. Second, this is basically the same mechanism we use for `@internal` functions, which worked well so far and makes code more uniform.

> And if we register the whole annotation in prelude, then having them as source files becomes rather pointless.
I don't think that those definitions in source files would be pointless. Those would ensure that for each annotation we have a source code position we could refer to if needed. Also many `@internal` functions are basically just [boring function declarations](https://github.com/dinfuehr/dora/blob/master/dora/stdlib/prelude.dora#L10-L17), not much difference to the annotations. You also mentioned parameters like `message` or `since` for `@deprecated`: those would still be defined in the source file.

Reply:
> So my assumption is that there needs to be some way to recognize these annotations in the runtime.

If you register them in the prelude, which purpose does the enum serve? Isn't `@internal` enough to check that the annotation is registered -- even though using an annotation to check semantics requires neither `InternalAnnotations`, nor the annotation marked `@internal`, nor having it in prelude (though convenient) -- you could define the annotation as source code and start using it in the compiler, no additional step is strictly necessary.

Reply:
> this InternalAnnotation enum seems more robust than using the Name returned by the interner

Yes, but holding the names in the interner was a pretty transient hack that wasn't supposed to survive namespaces anyway.

Reply:
Anyway, I'll register the annotations in prelude, move the modifier checks from the parser to the semantic phase, and hope for the best.

Perhaps a lightweight  `@target` is achievable, if the possible placements are defined as an enum, and we only check targets for the bare minimum required.

Reply:
> Yes, but holding the names in the interner was a pretty transient hack that wasn't supposed to survive namespaces anyway

Hmm, what would you do instead of holding the names in the future? My current understanding is that `@target` alone isn't enough to implement these internal annotations. For example how would the runtime find the `deprecated` annotation and its usages to provide warnings during compilation?

I don't really see the big difference between using the name or an enum. Both are just ways to find internal annotations. Yes, I think the latter one is more future proof and closer to the other instructions but it shouldn't change the design fundamentally. It should be no problem at all to introduce a method `Annotation::is_internal` which abstracts over both alternatives.

`@target` shouldn't even be affected at all. Internal annotations could definitely have a scope as well. However at least some internal annotations seem to require additional custom logic to verify correct usage, for example `@final` which can only be used on a virtual method from the parent class that is overwritten in the sub class.